File: mobile/lib/services/messaging/braze_provider.dart
import 'package:flutter/foundation.dart';

import '../service_providers.dart';

/// Braze messaging provider stub.
///
/// All methods print debug messages matching the ConsoleMessagingProvider pattern.
/// When `braze_plugin` is added as a dependency, replace stubs with real SDK calls.
class BrazeMessagingProvider extends MessagingProvider {
  @override
  String get name => 'braze';

  @override
  void init(Map<String, dynamic> config) {
    // Braze.instance.configure(apiKey: config['apiKey'], endpoint: config['endpoint'])
    debugPrint('[Braze] Initialized with config: ${config.keys.toList()}');
  }

  @override
  void setUser({
    required String userId,
    String? email,
    String? firstName,
    String? lastName,
    String? phone,
  }) {
    // braze.changeUser(userId)
    // braze.getUser().setEmail(email)
    // braze.getUser().setFirstName(firstName)
    // braze.getUser().setLastName(lastName)
    // braze.getUser().setPhoneNumber(phone)
    debugPrint('[Braze] setUser: $userId (email=$email)');
  }

  @override
  void clearUser() {
    // braze.wipeData()
    debugPrint('[Braze] clearUser');
  }

  @override
  void setUserAttributes(Map<String, dynamic> attributes) {
    // attributes.forEach((k, v) => braze.getUser().setCustomUserAttribute(k, v))
    debugPrint('[Braze] setUserAttributes: $attributes');
  }

  @override
  void logEvent(String event, {Map<String, dynamic>? properties}) {
    // braze.logCustomEvent(event, properties: properties)
    debugPrint('[Braze] logEvent: "$event" ${properties ?? {}}');
  }

  @override
  Future<bool> requestPushPermission() async {
    // Push permission is handled at the OS level; Braze registers the token automatically.
    // braze.requestPushPermission()
    debugPrint('[Braze] requestPushPermission (stub: auto-granted)');
    return true;
  }

  @override
  Future<bool> isPushEnabled() async {
    // braze.isPushNotificationEnabled()
    debugPrint('[Braze] isPushEnabled (stub: false)');
    return false;
  }

  @override
  void registerPushToken(String token) {
    // braze.registerAndroidPushToken(token) or handled automatically on iOS
    debugPrint('[Braze] registerPushToken: ${token.substring(0, 8)}...');
  }

  @override
  void handlePushReceived(Map<String, dynamic> payload) {
    // braze.handlePushNotification(payload)
    debugPrint('[Braze] handlePushReceived: $payload');
  }

  @override
  Future<void> flush() async {
    // braze.requestImmediateDataFlush()
    debugPrint('[Braze] flush');
  }

  @override
  Future<({bool granted, bool canRequest})> getPushPermissionStatus() async {
    // braze.getPushPermissionStatus()
    debugPrint('[Braze] getPushPermissionStatus (stub)');
    return (granted: false, canRequest: true);
  }

  @override
  void Function() onInAppMessage(void Function(Map<String, dynamic> message) callback) {
    // braze.subscribeToInAppMessages((message) => callback(message.toJson()))
    debugPrint('[Braze] onInAppMessage subscription registered (stub)');
    return () {
      debugPrint('[Braze] onInAppMessage subscription cancelled');
    };
  }

  @override
  void Function() onPushOpened(void Function(Map<String, dynamic> payload) callback) {
    // braze.subscribeToPushNotificationEvents((event) => callback(event.toJson()))
    debugPrint('[Braze] onPushOpened subscription registered (stub)');
    return () {
      debugPrint('[Braze] onPushOpened subscription cancelled');
    };
  }
}

File: apps/web/src/lib/services/messaging/braze-provider.ts
/**
 * Braze Messaging Provider
 *
 * Wraps the Braze Web SDK behind the MessagingProvider interface.
 * Requires `@braze/web-sdk` to be installed.
 *
 * Config:
 *   VITE_BRAZE_API_KEY — Braze API key
 *   VITE_BRAZE_SDK_ENDPOINT — Braze SDK endpoint (e.g. 'sdk.iad-01.braze.com')
 *
 * Install: npm install @braze/web-sdk
 */

import type {
  MessagingProvider,
  MessagingUser,
  InAppMessage,
  PushPermissionStatus,
} from "../types";

/** Minimal shape of the Braze Web SDK used by this provider */
interface BrazeLib {
  initialize(apiKey: string, options: Record<string, unknown>): void;
  openSession(): void;
  changeUser(userId: string): void;
  getUser(): {
    setEmail(email: string): void;
    setFirstName(name: string): void;
    setLastName(name: string): void;
    setPhoneNumber(phone: string): void;
    setCustomUserAttribute(key: string, value: unknown): void;
  };
  logCustomEvent(event: string, properties?: Record<string, unknown>): void;
  requestPushPermission(): void;
  isPushGranted(): boolean;
  subscribeToInAppMessage(
    callback: (msg: BrazeInAppMessage) => void,
  ): { removeSubscriber?: () => void } | undefined;
  subscribeToContentCardsUpdates(
    callback: (cards: unknown) => void,
  ): { removeSubscriber?: () => void } | undefined;
  showInAppMessage(msg: unknown): void;
  requestImmediateDataFlush(): void;
}

interface BrazeInAppMessage {
  messageId?: string;
  header?: string;
  message?: string;
  imageUrl?: string;
  extras?: Record<string, unknown>;
}

export class BrazeProvider implements MessagingProvider {
  readonly name = "braze";
  private braze: BrazeLib | null = null;

  async init(config: Record<string, unknown>): Promise<void> {
    const apiKey = config.apiKey as string;
    const sdkEndpoint = config.sdkEndpoint as string;

    if (!apiKey || !sdkEndpoint) {
      console.warn("[Braze] Missing apiKey or sdkEndpoint. Braze is disabled.");
      return;
    }

    try {
      // Dynamic import with variable indirection so Rollup/Vite skip static resolution
      const pkg = "@braze/web-sdk";
      const mod = await import(/* @vite-ignore */ pkg);
      const braze = mod.default ?? mod;
      braze.initialize(apiKey, {
        baseUrl: sdkEndpoint,
        enableLogging: config.debug === true,
        ...((config.options as Record<string, unknown>) ?? {}),
      });
      braze.openSession();
      this.braze = braze;
    } catch {
      console.warn("[Braze] @braze/web-sdk not installed. Run: npm install @braze/web-sdk");
    }
  }

  setUser(user: MessagingUser): void {
    if (!this.braze) return;
    this.braze.changeUser(user.userId);
    const brazeUser = this.braze.getUser();
    if (user.email) brazeUser.setEmail(user.email);
    if (user.firstName) brazeUser.setFirstName(user.firstName);
    if (user.lastName) brazeUser.setLastName(user.lastName);
    if (user.phone) brazeUser.setPhoneNumber(user.phone);
  }

  clearUser(): void {
    // Braze doesn't have a clearUser — changing to anonymous is the equivalent
    // A new session with no changeUser call is effectively anonymous
  }

  setUserAttributes(attributes: Record<string, unknown>): void {
    if (!this.braze) return;
    const brazeUser = this.braze.getUser();
    for (const [key, value] of Object.entries(attributes)) {
      brazeUser.setCustomUserAttribute(key, value);
    }
  }

  logEvent(event: string, properties?: Record<string, unknown>): void {
    if (!this.braze) return;
    this.braze.logCustomEvent(event, properties);
  }

  async requestPushPermission(): Promise<PushPermissionStatus> {
    if (!this.braze) return { granted: false, canRequest: false };
    try {
      this.braze.requestPushPermission();
      return { granted: true, canRequest: true };
    } catch {
      return { granted: false, canRequest: false };
    }
  }

  async getPushPermissionStatus(): Promise<PushPermissionStatus> {
    if (!this.braze) return { granted: false, canRequest: false };
    const isPushGranted = this.braze.isPushGranted();
    return { granted: isPushGranted, canRequest: !isPushGranted };
  }

  registerPushToken(_token: string): void {
    // Braze Web SDK handles push tokens automatically via service workers
  }

  handlePushReceived(_payload: Record<string, unknown>): void {
    // Braze handles push display via its service worker
  }

  onInAppMessage(callback: (message: InAppMessage) => void): { unsubscribe: () => void } {
    if (!this.braze) return { unsubscribe: () => {} };
    const sub = this.braze.subscribeToInAppMessage((msg: BrazeInAppMessage) => {
      callback({
        id: msg.messageId || String(Date.now()),
        title: msg.header,
        body: msg.message,
        imageUrl: msg.imageUrl,
        extras: msg.extras,
      });
      // Auto-display the message
      this.braze.showInAppMessage(msg);
    });
    return { unsubscribe: () => sub?.removeSubscriber?.() };
  }

  onPushOpened(callback: (payload: Record<string, unknown>) => void): { unsubscribe: () => void } {
    if (!this.braze) return { unsubscribe: () => {} };
    const sub = this.braze.subscribeToContentCardsUpdates((cards: unknown) => {
      callback({ cards });
    });
    return { unsubscribe: () => sub?.removeSubscriber?.() };
  }

  async flush(): Promise<void> {
    if (!this.braze) return;
    this.braze.requestImmediateDataFlush();
  }
}
